### ISSUE TITLE
`in` operator: check key/property is valid (using `keyof`)

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#50953]
# Suggestion

## 🔍 Search Terms

- in operator
- keyof
- key
- property
- validate

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

When using the `in` operator `K in T`, TypeScript should error if the provided key `K` is not assignable to `keyof T`.

## 📃 Motivating Example

```ts
const obj = { a: "", b: "" };

// This condition will always be false
const condition = "c" in obj;
```

```ts
declare const obj:
    | {
          name: string;
          age: string;
      }
    | {
          species: string;
      };

// This is a mistake, but it's not caught by the compiler.
const condition = 'nam' in obj;
```

## 💻 Use Cases

> What shortcomings exist with current approaches?

I've tried to define a `has` function for this purpose but there are lots of edge cases: https://stackoverflow.com/questions/58749852/typescript-defining-the-in-operator-as-a-function

**Comment by MartinJohns** [comment]
Note that the `in` operator is explicitly mentioned as a way to narrow union types in the handbook: https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing

**Comment by RyanCavanaugh** [comment]
This is opposite to the highly-requested behavior requested in #21732, which we just implemented

```ts
const obj = { a: "", b: "" };
if("c" in obj) {
  // Works in 4.9 nightly
  obj.c
}
```

With exact types we could revisit, but we don't have exact types

**Comment by OliverJAsh** [comment]
@MartinJohns This suggestion isn't going against that, as far as I can see. The narrowing is useful. I just want TS to check that it's a valid key in at least one of the union members, so I can't make silly mistakes.

@RyanCavanaugh I think #21732 is really useful, when we're dealing with a type like `unknown` or `object`, but in this case we already have type information about the properties (assuming that the type is exact like you say).

**Comment by RyanCavanaugh** [comment]
If the object isn't aliased in some way, why are you writing an `in` check?

**Comment by OliverJAsh** [comment]
Here's one example:

```ts
type Fish = { swim: () => void };
type Bird = { fly: () => void };
type Human = { swim?: () => void; fly?: () => void };

const getType = (animal: Fish | Bird | Human) => {
    // oops, typo!
    if ("flyyy" in animal) {
        return "bird";
    } else {
        return "other";
    }
};

```

**Comment by MartinJohns** [comment]
> The narrowing is useful. I just want TS to check that it's a valid key in at least one of the union members, so I can't make silly mistakes.

You explicitly mentioned `keyof T`, which only includes keys common in all union types.

**Comment by OliverJAsh** [comment]
You're right. I should have said something like "distributive `keyof`" (`type DistributiveKeys<T> = T extends any ? keyof T : never`).

**Comment by RyanCavanaugh** [comment]
> I think https://github.com/microsoft/TypeScript/issues/21732 is really useful, when we're dealing with a type like unknown or object [...]

Sure, but what you're proposing breaks subtyping linearity: if `k in T` is valid when `T` is `unknown` (or `object`), it doesn't make sense to say that `k in T'` is invalid when `T'` is some subtype of `T` (which, for unknown, *is* by definition).  There are some places where breaking that linearity makes sense, but they're pretty few and far between IMO.

If you had written
```ts
type Fish = { swim: () => void };
type Bird = { fly: () => void };
type Human = { swim?: () => void; fly?: () => void };

const getType = (animal: Fish | Bird | Human) => {
    if ("superman" in animal) {
        return "kryptonian";
    } else {
        return "earthling";
    }
};
```
there doesn't seem to be an error in this program.

**Comment by MartinJohns** [comment]
@OliverJAsh You could probably easily write an ESLint rule for this, if you still want it.

**Comment by LexRiver** [comment]
> there doesn't seem to be an error in this program.

Can you explain why?
In Javascript there is no error, but in Typescript we expecting only types ```Fish | Bird | Human```, not an extensions of these types, so we must work with known properties of these types.
If we expecting superman in this function we should write ```animal: Fish | Bird | Human | Superman```

**Comment by RyanCavanaugh** [comment]
Assuming we say "Superman is a human" (paging Plato?), the union `Fish | Bird | Human | Superman` is * set equivalent* to `Fish | Bird | Human`, so it doesn't make sense to say you should have to write that.

**Comment by LexRiver** [comment]
Ok here is another syntax proposal:

```typescript
type A = {
    a1:string
    a2:number
    a3?:boolean
}

interface B {
    b1:string
    b2?:number
    b3?:boolean
}

function func(x:A|B){
    if(x satisfies A){ // compiles to: if('a1' in x && 'a2' in x){ // we check all non-optional properties of type
        console.log('A')
    }
    if(x satisfies B){ // compiles to: if('b1' in x){
        console.log('B')
    }
}

const a:A = {
    a1: 'a1',
    a2: 2,
}

const b:B = {
    b1: 'b1',
    b2: 2,
    b3: true
}

func(a) // output: A
func(b) // output: B
func({a1:'a1', b1:'b1'}) // output: B
func({a1:'a1', a2:2, b1:'b1'}) // output: A B
```

instead of `satisfies` could be some another keyword.

**Comment by RyanCavanaugh** [comment]
That's very explicitly out-of-scope for TS

**Comment by LexRiver** [comment]
how about this? -)
```typescript
    if('a1' as keyof A in x){
        console.log('A')
    }
    if('b1' as keyof A in x){ // should be error
        //
    }

```

**Comment by RyanCavanaugh** [comment]
You can already write
```ts
    if('b1' satisfies keyof A in x){ // should be error
        //
    }
```
to get an error in that situation

**Comment by LexRiver** [comment]
Great, I didn't know that.
I think this should be mentioned in documentation
https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing
because current approach with `in` narrowing leads to a bugs in a long term.
